### DISCUSSION TITLE
How to use local cache result as remote cache?

### DISCUSSION THREAD

**Discussion by fanelena** [ccache/ccache#1220] [Questions & answers] [answered]
Hello :)
The ccache provides a lot of usefulness.

I am using ccache version 4.7.
What I want is sharing the primary cache stored on the local machine by moving to https server for sharing.
When I check it, primary cache file path and name is some different with seconday storage(remote storage).

ex)
Hash 9957j3unq5ma477ljoqia8fncbbtm1cru is stored
Primary:
9/9/57j3unq5ma477ljoqia8fncbbtm1cruR
Secondary:
99/57j3unq5ma477ljoqia8fncbbtm1cru

On manual, layout option exist but All "flat", "subdirs" do not matched with primary layout.
Yes, I can simply convert one to another with shell script.
Is ccache support option to maching these two?

Thanks in advance :)

**Comment by afbjorklund** [comment, marked as answer]
Currently there are only some helper scripts in python to upload and download from a **redis** cache:

https://github.com/ccache/ccache/tree/master/misc

One idea was to convert them to c++ and to make them more generic, but so far no volunteers...

I did make some similar scripts to convert to and from .tar format, that _could_ be used for http too ?

* https://github.com/ccache/ccache/issues/1118 ([`create-archive`](https://github.com/afbjorklund/ccache/blob/ce1f7efb3c4b66a1507b461eb5fda2c71807eaf2/misc/create-archive))

The "ccache.tar" would be unpacked in the www root, after copying the generated file to the web server.

> Is ccache support option to maching these two?

No, there is no such ccache option as far as I know.

**Reply by fanelena** [reply]
Thanks for sharing your script :)

**Reply by afbjorklund** [reply]
Note: the files are sorted, so it should be possible to do delta updates (using rsync) if your cache is large

`rsync --info=progress2 ccache.tar cache.example.com:/var/tmp/ccache.tar`

Did some other crazy experiments, but in the end you need to unpack/mount the archive before using it...

Skipping http(s) and using redis with one of the [file-backed redis options](https://github.com/ccache/ccache/wiki/Redis-storage) also works.

**Comment by jrosdahl** [comment]
> The ccache provides a lot of usefulness.

Thanks!

> [...]
> On manual, layout option exist but All "flat", "subdirs" do not matched with primary layout. Yes, I can simply convert one to another with shell script. Is ccache support option to maching these two?

From your question, I'm not sure if (A) you just want to transfer the local storage data to another server as a one-time effort, or if (B) you want make the local data available via HTTP on the local host.

Answer to A: There is no official way to do that and you'll have to use some custom script. Note that if you use [hard linking](https://ccache.dev/manual/4.7.3.html#config_hard_link) or [file cloning](https://ccache.dev/manual/4.7.3.html#config_file_clone) you can't use such results on the HTTP server.

Answer to B: The local storage can't be used as the backing store of remote storage, and that's intentional. There are three reasons why local and remote storage can't use the same directory:

1. When ccache writes to local storage, it also maintains metadata about things like cache size. If remote storage would write files to the same file structure then that metadata would become out of sync. Remote storage is just a "dumb" key-value store that maps keys to a static file structure; it doesn't have any metadata or other bookkeeping.
2. The local storage has a dynamic number of directory levels. An `abcdef` entry is by default stored as `a/b/cdef`. When the number of files in `a/b` reaches a certain size, the entry will be stored as `a/b/c/def` instead, etc. This behavior depends on the metadata mentioned above which is available in local storage but not in remote storage.
3. If you have enabled [hard linking](https://ccache.dev/manual/4.7.3.html#config_hard_link) or [file cloning](https://ccache.dev/manual/4.7.3.html#config_file_clone), local entries cannot be used by a remote backend.

What you can do, though, is to use [remote file storage](https://ccache.dev/manual/4.7.3.html#_file_storage_backend) in combination with [`remote_only = true`](https://ccache.dev/manual/4.7.3.html#config_remote_only) and configure file and HTTP storage to use the same directory.

That is, use configuration like this locally:

```
remote_only = true
remote_storage = file:///example/cache
```

Then set up an HTTP server that serves `/example/cache` on, say, `http://host/cache` and use `remote_storage = http://host/cache` from other machines.

**Reply by fanelena** [reply]
Thanks for the detailed comment.
What I want to do is close to case "A".
For general developers, I want to limit the write permission to the remote storage and daily official build result is only updated to
server. I assumed that I don't have permission to directly write to http or nfs server.
I set remote_storage option as "read-only|share-hits=false" to store only missed result to my local. That's for update partial differetns to server through rsync.

Another my considered configuration was that setting secondary storage as my local file path. And use it as for sharing in server.
ex) CCACHE_SECONDARY_STORAGE="file:///${HOME}/ccache_ext"
